---
title: "Script for extracting images"
author: "Imaging Group 8"
#subtitle
#date
params:
  soln: TRUE   ## EDIT to TRUE when generating solution, otherwise 'FALSE'
  supp: FALSE
  show: 'as.is'  ## EDIT to 'as.is' when generating Suggestions, otherwise 'hide'
output:
  html_document:
    fig_caption: yes
    include:
      after_body: css/stylesDD.js
    number_sections: yes
    embed-resources: true
    theme: flatly
    css: 
      - css/styles.css
      - https://use.fontawesome.com/releases/v5.0.6/css/all.css
    toc: true
    toc_depth: 3
    toc_float: true
    code_folding: hide
  pdf_document:
    number_sections: yes
    toc: yes
---

<br><br>

<div class="aimbox"> 
### <span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-pencil-alt fa-stack-1x fa-inverse"></i></span> Document summary {-}

- This RMarkdown document serves as a reproducible code for generating the Biotechnology per-cell images.
- This can be fully *re-run* assuming:
  - Similar directory structure as described below, and
  - Mac/linux operating system.

### <span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-location-arrow fa-stack-1x fa-inverse"></i></span> Key considerations {-}
- Only a randomly selected subset of 1,000 cells' images are extracted in this document.
- Cell labels are the result of Graph-based clustering (an *unsupervised* learning technique) of *gene expression* data extracted for these cells. Can this be thought of as "ground truth"?
<br>
</div>
<br>

# Full data description

The data described in this document stems from a new biotechnology: 
molecule-resolved spatial genomics. In particular we will explore data that has 
been generated by [10x Genomics](https://www.10xgenomics.com/) Xenium 
instrument on a fresh frozen mouse brain coronal section - Tiny subset. The technology
results in several outputs including:

- cell morphology image where intensity corresponds to presence of the nucleus 
of each cell.
- cell boundaries indicating spatial locations of detected cells
- RNA abundances (gene expression) of each cell, which has been grouped into
28 distinct clusters, of which cluster labels are provided.

The full data and description can be found in this [link](https://www.10xgenomics.com/resources/datasets/fresh-frozen-mouse-brain-for-xenium-explorer-demo-1-standard).

# Data and code directory structure

For this reproducible code we assume that this RMarkdown document is saved within the following directory structure:

```
├── data/
│   ├── data_processed/
│   ├── cell_boundaries.csv.gz
│   ├── clusters.csv
│   ├── differential_expression.csv
│   ├── Gene Groups.csv
│   ├── morphology_focus.tif
│   └── scripts/
│       └── image_extract_script.Rmd (this script)
└── shiny_app/
```
    
All the following files can be retrieved from the data given from the lab, except for Gene Groups.csv and differential_expression.csv

The gene_group csv can be retrieved here: https://www.10xgenomics.com/datasets/fresh-frozen-mouse-brain-for-xenium-explorer-demo-1-standard

The contents of `data_raw/` has been removed due to the large file size.

# The `EBImage` package

`EBImage` is an R package that is available in the 
[Bioconductor Project](http://bioconductor.org/).
Bioconductor is similar to the 
[Comprehensive R Archive Network (CRAN)](https://cran.r-project.org/), 
in that you can install packages from this repository.

[`EBImage`](https://bioconductor.org/packages/release/bioc/html/EBImage.html) 
provides general purpose functionality for image processing and analysis. In 
the context of (high-throughput) microscopy-based cellular assays, EBImage 
offers tools to segment cells and extract quantitative cellular descriptors. 
This allows the automation of such tasks using the R programming language and 
facilitates the use of other tools in the R environment for signal processing, 
statistical modeling, machine learning and visualization with image data.

This [chapter in Modern Statistics for Modern Biology](https://web.stanford.edu/class/bios221/book/11-chap.html) is a great reference for using `EBImage` for different types of imaging data.

To install the `EBImage` package, you can run the chunk below. This will 
check whether you have the `BiocManager` package installed, and if not it will
install `BiocManager`. Then, the `EBImage` package will be installed via
the `BiocManager::install()` function.

Note: if you attempt to run `install.packages("EBImage")` you may be met with 
an error! This is because the package is available in Bioconductor and not on 
CRAN.

```
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("EBImage")
```

```{r, include=FALSE}
knitr::opts_chunk$set(
    warning = FALSE, message = FALSE
)
```


```{r, warning=FALSE, message=FALSE}
library(tidyverse)
library(EBImage)
library(dplyr)
library(pracma)
library(png)
```



```{r}
get_inside = function(cellID, img, cell_boundaries) {
  
  cell_boundary = cell_boundaries |>
    filter(cell_id %in% cellID)
  
  # rescale the boundary according to the pixels
  pixels = dim(img)
  cell_boundary$vertex_x_scaled <- 1+((cell_boundary$vertex_x - min(cell_boundary$vertex_x))/0.2125)
  cell_boundary$vertex_y_scaled <- 1+((cell_boundary$vertex_y - min(cell_boundary$vertex_y))/0.2125)
  
  # identify which pixels are inside or outside of the cell segment using inpolygon
  pixel_locations = expand.grid(seq_len(nrow(img)), seq_len(ncol(img)))
  
  pixels_inside = inpolygon(x = pixel_locations[,1],
                            y = pixel_locations[,2],
                            xp = cell_boundary$vertex_x_scaled,
                            yp = cell_boundary$vertex_y_scaled,
                            boundary = TRUE)
  
  img_inside = img
  img_inside@.Data <- matrix(pixels_inside, nrow = nrow(img), ncol = ncol(img))
  
  return(img_inside)
}

mask_resize = function(img, img_inside, w = 50, h = 50) {
  
  img_mask = img*img_inside
  
  # then, transform the masked image to the same number of pixels, 50x50
  img_mask_resized = resize(img_mask, w, h)
  
  return(img_mask_resized)
}

save_as_png <- function(img, filename) {
  png::writePNG(img, filename)
}

```

# Read morphology image

Note: If you did not have the tif file here, head over to the raw_image_process.Rmd file for instructions.

```{r}
tifFile = "../morphology_focus.tif"
if (!file.exists(tifFile)) {
  options(java.parameters = "-Xmx10g")
  library(RBioFormats)
  checkJavaMemory()
  img_ome = RBioFormats::read.image("../data_raw/morphology_focus.ome.tif",
                                    read.metadata = FALSE,
                                    normalize = TRUE)
  
  img = img_ome[[1]]@.Data
  EBImage::writeImage(x = img,
                      files = tifFile,
                      type = "tiff")
}
img = EBImage::readImage(tifFile)

```

# Labelling gene by gene group

```{r}
diff_exp = read.csv("../differential_expression.csv")
n_clusters = ncol(diff_exp |> select(contains("Adjusted.p.value")))
diff_exp <- diff_exp |> select(contains(c("Feature.Name","Adjusted.p.value","fold")))
colnames(diff_exp)[1] <- "gene"
gene_groups = read.csv("../Gene Groups.csv")
gene_groups <- gene_groups |> mutate(group = ifelse(group == "", "Emp", group))

diff_exp <- left_join(diff_exp, gene_groups, "gene")

diff_exp <- diff_exp |> replace_na(list(group = "Emp"))
```

```{r}
unique(diff_exp$group)
```

# Decide group by differentially expressed genes, subset 500 images each class

```{r, warning = FALSE}
cell_boundaries = read.csv("../cell_boundaries.csv.gz")
cell_boundaries$vertex_x_trans = cell_boundaries$vertex_x/0.2125
cell_boundaries$vertex_y_trans = cell_boundaries$vertex_y/0.2125
clusters = read.csv("../clusters.csv")
ncells = nrow(clusters)

for (i in 1:n_clusters){
    lowest_ps = which(diff_exp[, i + 1] <= quantile(diff_exp[, i + 1], 0.00001))
    max_fc_index = lowest_ps[which.max(diff_exp[lowest_ps, i + n_clusters + 1])]
    label = diff_exp[max_fc_index, 2 * n_clusters + 2]
    clusters <- clusters |> mutate(Cluster = ifelse(Cluster == i, label, Cluster))
}



emp_id = clusters %>% filter(Cluster == "Emp") %>% pull(Barcode)
astro_id = clusters %>% filter(Cluster == "Astrocytes") %>% pull(Barcode)
endo_id =  clusters %>% filter(Cluster == "Endothelial") %>% pull(Barcode)
mic_id = clusters %>% filter(Cluster == "Microglia") %>% pull(Barcode)
neuron_id = clusters %>% filter(Cluster == "Neurons") %>% pull(Barcode)
oli_id = clusters %>% filter(Cluster == "Oligodendrocytes") %>% pull(Barcode)

set.seed(2024)

# Sample 500 IDs from each vector
sample_emp <- sample(emp_id, 500, replace = FALSE)
sample_astro <- sample(astro_id, 500, replace = FALSE)
sample_endo <- sample(endo_id, 500, replace = FALSE)
sample_mic <- sample(mic_id, 500, replace = FALSE)
sample_neuron <- sample(neuron_id, 500, replace = FALSE)
sample_oli <- sample(oli_id, 500, replace = FALSE)

# Combine all sampled IDs into a single vector
cells_subset <- c(sample_emp, sample_astro, sample_endo, sample_mic, sample_neuron, sample_oli)


#ncells_subset = 1000

#cells_subset = sample(ncells,ncells_subset)

for (i in cells_subset) {
  
  # extract the boundary vertices for the selected cell
  bounds_i = subset(cell_boundaries,cell_id == i)
  
  # extract the cluster value for the selected cell
  clustval_i = clusters %>% filter(Barcode == i) %>% select(Cluster) %>% as.character()
  
  # extract the pixel intensities for the area covering the cell boundary
  img_sub = img[min(bounds_i$vertex_x_trans):max(bounds_i$vertex_x_trans),
                min(bounds_i$vertex_y_trans):max(bounds_i$vertex_y_trans)]
  
  # normalise the pixel intensities according to 99th percentile
  img_sub_norm = img_sub/quantile(img_sub, 0.99)
  
  # if there is no folder for cell_images, create one
  if (!file.exists("../data_processed/cell_images_regrouped/")) {
    dir.create("../data_processed/cell_images_regrouped/", recursive = TRUE, showWarnings = FALSE)
  }
  
  # if there is no folder for the cluster, create one
  clustval_i_directory = paste0("../data_processed/cell_images_regrouped/",clustval_i)
  if (!file.exists(clustval_i_directory)) {
    dir.create(clustval_i_directory, recursive = TRUE, showWarnings = FALSE)
  }
  
  # save the extracted image as a png file
  EBImage::writeImage(x = img_sub_norm,
                      files = paste0(clustval_i_directory, "/cell_", i, ".png"),
                      type = "png")
  
}
```

# Resizing balanced image set

```{r}
i = 1
groups = list.files("../data_processed/cell_images_regrouped",full.names = FALSE)
while(i <= 6){
  input_dir = paste("../data_processed/cell_images_regrouped/",groups[i],sep="")
  cell_files = list.files(input_dir,full.names = TRUE)
  cell_images = sapply(cell_files, readImage, simplify = FALSE)
  cell_ids = gsub(".*cell_|.png", "", cell_files)
  imgs_inside = mapply(get_inside, cell_ids, cell_images, MoreArgs = list(cell_boundaries = cell_boundaries), SIMPLIFY = FALSE)
  cell_images_masked_resized = mapply(mask_resize, cell_images, imgs_inside, SIMPLIFY = FALSE)
  output_dir = paste("../data_processed/cells_renamed_balanced/",groups[i],"/",sep="")
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  for (k in seq_along(cell_images_masked_resized)) {
    filename <- paste0(output_dir, cell_ids[k],".png")
    save_as_png(cell_images_masked_resized[[k]], filename)
  }
  i = i + 1
}

```

# Subset 1000 images

```{r}
ncells_subset = 1000

cells_subset = sample(clusters$Barcode,ncells_subset)

for (i in cells_subset) {
  
  # extract the boundary vertices for the selected cell
  bounds_i = subset(cell_boundaries,cell_id == i)
  
  # extract the cluster value for the selected cell
  clustval_i = clusters %>% filter(Barcode == i) %>% select(Cluster) %>% as.character()
  
  # extract the pixel intensities for the area covering the cell boundary
  img_sub = img[min(bounds_i$vertex_x_trans):max(bounds_i$vertex_x_trans),
                min(bounds_i$vertex_y_trans):max(bounds_i$vertex_y_trans)]
  
  # normalise the pixel intensities according to 99th percentile
  img_sub_norm = img_sub/quantile(img_sub, 0.99)
  
  # if there is no folder for cell_images, create one
  if (!file.exists("../data_processed/cell_images_regrouped_2/")) {
    dir.create("../data_processed/cell_images_regrouped_2/", recursive = TRUE, showWarnings = FALSE)
  }
  
  # if there is no folder for the cluster, create one
  clustval_i_directory = paste0("../data_processed/cell_images_regrouped_2/",clustval_i)
  if (!file.exists(clustval_i_directory)) {
    dir.create(clustval_i_directory, recursive = TRUE, showWarnings = FALSE)
  }
  
  # save the extracted image as a png file
  EBImage::writeImage(x = img_sub_norm,
                      files = paste0(clustval_i_directory, "/cell_", i, ".png"),
                      type = "png")
  
}
```

# Resize and masking imbalanced image set.

```{r}
i = 1
groups = list.files("../data_processed/cell_images_regrouped_2",full.names = FALSE)
while(i <= 6){
  input_dir = paste("../data_processed/cell_images_regrouped_2/",groups[i],sep="")
  cell_files = list.files(input_dir,full.names = TRUE)
  cell_images = sapply(cell_files, readImage, simplify = FALSE)
  cell_ids = gsub(".*cell_|.png", "", cell_files)
  imgs_inside = mapply(get_inside, cell_ids, cell_images, MoreArgs = list(cell_boundaries = cell_boundaries), SIMPLIFY = FALSE)
  cell_images_masked_resized = mapply(mask_resize, cell_images, imgs_inside, SIMPLIFY = FALSE)
  output_dir = paste("../data_processed/cells_renamed_imbalanced/",groups[i],"/",sep="")
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  for (k in seq_along(cell_images_masked_resized)) {
    filename <- paste0(output_dir, cell_ids[k],".png")
    save_as_png(cell_images_masked_resized[[k]], filename)
  }
  i = i + 1
}
```

# Plot distribution of image data after relabelling

```{r}
count_per_group <- clusters |> mutate(Cluster = ifelse(Cluster == "Emp", "Unclassified", Cluster))

p1 <- ggplot(count_per_group, aes(x = Cluster, fill = Cluster)) +
    geom_bar() + scale_fill_viridis_d() + theme_bw()
ggsave(width = 8, height = 4.5, "../../figures/distribution_plot.png", plot = p1)
```

# Saving image set to R objects

```{r}
astrocyte_files <- list.files("../data_processed/cell_images_regrouped/Astrocytes/", full.names = TRUE)

empty_files <- list.files("../data_processed/cell_images_regrouped/Emp/", full.names = TRUE)

endothelial_files <- list.files("../data_processed/cell_images_regrouped/Endothelial/", full.names = TRUE)

microglia_files <- list.files("../data_processed/cell_images_regrouped/Microglia/", full.names = TRUE)

neuron_files <- list.files("../data_processed/cell_images_regrouped/Neurons/", full.names = TRUE)

oligodendrocyte_files <- list.files("../data_processed/cell_images_regrouped/Oligodendrocytes/", full.names = TRUE)

astrocyte_files_2 <- list.files("../data_processed/cells_renamed_balanced/Astrocytes/", full.names = TRUE)

empty_files_2 <- list.files("../data_processed/cells_renamed_balanced/Emp/", full.names = TRUE)

endothelial_files_2 <- list.files("../data_processed/cells_renamed_balanced/Endothelial/", full.names = TRUE)

microglia_files_2 <- list.files("../data_processed/cells_renamed_balanced/Microglia/", full.names = TRUE)

neuron_files_2 <- list.files("../data_processed/cells_renamed_balanced/Neurons/", full.names = TRUE)

oligodendrocyte_files_2 <- list.files("../data_processed/cells_renamed_balanced/Oligodendrocytes/", full.names = TRUE)


all_files <- list.files("../data_processed/cells_renamed_balanced/", full.names = TRUE,recursive = TRUE)
```


```{r}
astro_cell_ids = gsub(".*cell_|.png", "", astrocyte_files)
empty_cell_ids = gsub(".*cell_|.png", "", empty_files)
endo_cell_ids = gsub(".*cell_|.png", "", endothelial_files)
micro_cell_ids = gsub(".*cell_|.png", "", microglia_files)
neuron_cell_ids = gsub(".*cell_|.png", "", neuron_files)
oligo_cell_ids = gsub(".*cell_|.png", "", oligodendrocyte_files)

astro_cell_ids_2 = gsub(".*/|.png", "", astrocyte_files_2)
empty_cell_ids_2 = gsub(".*/|.png", "", empty_files_2)
endo_cell_ids_2 = gsub(".*/|.png", "", endothelial_files_2)
micro_cell_ids_2 = gsub(".*/|.png", "", microglia_files_2)
neuron_cell_ids_2 = gsub(".*/|.png", "", neuron_files_2)
oligo_cell_ids_2 = gsub(".*/|.png", "", oligodendrocyte_files_2)

all_cell_ids = gsub(".*/|.png", "", all_files)

```

```{r}
astro_imgs <- sapply(astrocyte_files, readImage, simplify = FALSE)
names(astro_imgs) <- astro_cell_ids

empty_imgs <- sapply(empty_files, readImage, simplify = FALSE)
names(empty_imgs) <- empty_cell_ids

endothelial_imgs <- sapply(endothelial_files, readImage, simplify = FALSE)
names(endothelial_imgs) <- endo_cell_ids

microglia_imgs <- sapply(microglia_files, readImage, simplify = FALSE)
names(microglia_imgs) <- micro_cell_ids

neuron_imgs <- sapply(neuron_files, readImage, simplify = FALSE)
names(neuron_imgs) <- neuron_cell_ids

oligodendrocyte_imgs <- sapply(oligodendrocyte_files, readImage, simplify = FALSE)
names(oligodendrocyte_imgs) <- oligo_cell_ids

astro_imgs_2 <- sapply(astrocyte_files_2, readImage, simplify = FALSE)
names(astro_imgs_2) <- astro_cell_ids_2

empty_imgs_2 <- sapply(empty_files_2, readImage, simplify = FALSE)
names(empty_imgs_2) <- empty_cell_ids_2

endothelial_imgs_2 <- sapply(endothelial_files_2, readImage, simplify = FALSE)
names(endothelial_imgs_2) <- endo_cell_ids_2

microglia_imgs_2 <- sapply(microglia_files_2, readImage, simplify = FALSE)
names(microglia_imgs_2) <- micro_cell_ids_2

neuron_imgs_2 <- sapply(neuron_files_2, readImage, simplify = FALSE)
names(neuron_imgs_2) <- neuron_cell_ids_2

oligodendrocyte_imgs_2 <- sapply(oligodendrocyte_files_2, readImage, simplify = FALSE)
names(oligodendrocyte_imgs_2) <- oligo_cell_ids_2

all_cell_imgs <- sapply(all_files, readImage, simplify = FALSE)
names(all_cell_imgs) <- all_cell_ids
```

```{r}
#|eval: false

# if there is no folder for cell_images, create one
if (!file.exists("../data_processed/images_R_obj/")) {
    dir.create("../data_processed/images_R_obj/", recursive = TRUE, showWarnings = FALSE)
}

if (!file.exists("../../shiny_app/images_R_obj/")) {
    dir.create("../../shiny_app/images_R_obj/", recursive = TRUE, showWarnings = FALSE)
}


saveRDS(astro_imgs, "../data_processed/images_R_obj/astro.rds")
saveRDS(empty_imgs, "../data_processed/images_R_obj/empty.rds")
saveRDS(endothelial_imgs, "../data_processed/images_R_obj/endo.rds")
saveRDS(microglia_imgs, "../data_processed/images_R_obj/micro.rds")
saveRDS(neuron_imgs, "../data_processed/images_R_obj/neuron.rds")
saveRDS(oligodendrocyte_imgs, "../data_processed/images_R_obj/oligo.rds")

saveRDS(astro_imgs_2, "../data_processed/images_R_obj/astro_2.rds")
saveRDS(empty_imgs_2, "../data_processed/images_R_obj/empty_2.rds")
saveRDS(endothelial_imgs_2, "../data_processed/images_R_obj/endo_2.rds")
saveRDS(microglia_imgs_2, "../data_processed/images_R_obj/micro_2.rds")
saveRDS(neuron_imgs_2, "../data_processed/images_R_obj/neuron_2.rds")
saveRDS(oligodendrocyte_imgs_2, "../data_processed/images_R_obj/oligo_2.rds")

saveRDS(all_cell_imgs, "../data_processed/images_R_obj/all.rds")


saveRDS(astro_imgs, "../../shiny_app/images_R_obj/astro.rds")
saveRDS(empty_imgs, "../../shiny_app/images_R_obj/empty.rds")
saveRDS(endothelial_imgs, "../../shiny_app/images_R_obj/endo.rds")
saveRDS(microglia_imgs, "../../shiny_app/images_R_obj/micro.rds")
saveRDS(neuron_imgs, "../../shiny_app/images_R_obj/neuron.rds")
saveRDS(oligodendrocyte_imgs, "../../shiny_app/images_R_obj/oligo.rds")

saveRDS(astro_imgs_2, "../../shiny_app/images_R_obj/astro_2.rds")
saveRDS(empty_imgs_2, "../../shiny_app/images_R_obj/empty_2.rds")
saveRDS(endothelial_imgs_2, "../../shiny_app/images_R_obj/endo_2.rds")
saveRDS(microglia_imgs_2, "../../shiny_app/images_R_obj/micro_2.rds")
saveRDS(neuron_imgs_2, "../../shiny_app/images_R_obj/neuron_2.rds")
saveRDS(oligodendrocyte_imgs_2, "../../shiny_app/images_R_obj/oligo_2.rds")

saveRDS(all_cell_imgs, "../../shiny_app/images_R_obj/all.rds")

```


# Finish

```{r}
sessionInfo()
```

```{r}
knitr::knit_exit()
```