---
title: "Script for Processing Raw files"
author: "Imaging Group 8"
#subtitle
#date
params:
  soln: TRUE   ## EDIT to TRUE when generating solution, otherwise 'FALSE'
  supp: FALSE
  show: 'as.is'  ## EDIT to 'as.is' when generating Suggestions, otherwise 'hide'
output:
  html_document:
    fig_caption: yes
    include:
      after_body: css/stylesDD.js
    number_sections: yes
    embed-resources: true
    theme: flatly
    css: 
      - css/styles.css
      - https://use.fontawesome.com/releases/v5.0.6/css/all.css
    toc: true
    toc_depth: 3
    toc_float: true
    code_folding: hide
  pdf_document:
    number_sections: yes
    toc: yes
---

```{r initialsetup, include=FALSE}
# For suggestions
chooseCRANmirror(graphics=FALSE, ind=1)
knitr::opts_chunk$set(echo = TRUE,tidy=TRUE)
show_q = FALSE
show_s = FALSE
# Then include suggestions file
# ```{r child='Lab1_Quiz_s.Rmd', eval = show_s}
```

<br><br>

<div class="aimbox"> 
### <span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-pencil-alt fa-stack-1x fa-inverse"></i></span> Document summary {-}

- This RMarkdown document serves as a reproducible code for generating the Biotechnology per-cell images.
- This can be fully *re-run* assuming:
  - Enough disk capacity to download full Biotechnology data bundle,
  - Enough memory to read `ome.tif` image via `RBioFormats` package,
  - Similar directory structure as described below, and
  - Mac/linux operating system.
- The *output* is the zipped file `Biotechnology.zip`, which has already been provided on Canvas.

### <span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-location-arrow fa-stack-1x fa-inverse"></i></span> Key considerations {-}
- Only a randomly selected subset of 1,000 cells' images are extracted in this document.
- Cell labels are the result of Graph-based clustering (an *unsupervised* learning technique) of *gene expression* data extracted for these cells. Can this be thought of as "ground truth"?
<br>
</div>
<br>

# Full data description

The data described in this document stems from a new biotechnology: 
molecule-resolved spatial genomics. In particular we will explore data that has 
been generated by [10x Genomics](https://www.10xgenomics.com/) Xenium 
instrument on a fresh frozen mouse brain coronal section - Tiny subset. The technology
results in several outputs including:

- cell morphology image where intensity corresponds to presence of the nucleus 
of each cell.
- cell boundaries indicating spatial locations of detected cells
- RNA abundances (gene expression) of each cell, which has been grouped into
28 distinct clusters, of which cluster labels are provided.

The full data and description can be found in this [link](https://www.10xgenomics.com/resources/datasets/fresh-frozen-mouse-brain-for-xenium-explorer-demo-1-standard).

# Data and code directory structure

For this reproducible code we assume that this RMarkdown document is saved within the following directory structure:

```
├── data/
    ├── cell_boundaries.csv.gz
    ├── clusters.csv
    ├── data_processed/
    ├── data_raw/
    │   ├── <extracted files>/
    │   └── Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP_outs.zip (3.5Gb zip file)
    ├── differential_expression.csv
    ├── Gene Groups.csv (get from website)
    ├── morphology_focus.tif
    └── scripts/
        └── DATA3888_Biotechnology_generateImages_2024.Rmd (this script)
```
    
For you to be able to fully re-run this code you will need to download the
contents of `data_raw/` separately (see next section).

You are provided this directory structure, with the contents of `data_raw/`
removed due to the large file size.

# Raw data bundle

The contents to `Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP_outs` folder is 
from the `Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP_outs.zip` file (approx 3.5GB), available
to download via this [LINK](https://cf.10xgenomics.com/samples/xenium/1.0.2/Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP/Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP_outs.zip), or can be programmatically downloaded
using `wget` into the target directory.

```
wget https://cf.10xgenomics.com/samples/xenium/1.0.2/Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP/Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP_outs.zip ../data_raw/
unzip ../data_raw/Xenium_V1_FF_Mouse_Brain_Coronal_Subset_CTX_HP_outs.zip -d ../data_raw/
```

Also, you can download suplimentary data Gene Groups.csv here: 

```
curl -O https://cf.10xgenomics.com/samples/xenium/1.0.2/Xenium_V1_FF_Mouse_Brain_Coronal_Input/Xenium_V1_FF_Mouse_Brain_Coronal_Input_gene_groups.csv

mv Xenium_V1_FF_Mouse_Brain_Coronal_Input_gene_groups.csv 'Gene Groups.csv'
```


Note! It is very important to ensure you are working from the *correct
working directory*, i.e. within the `scripts` folder in the directory structure 
described above.

# The `EBImage` package

`EBImage` is an R package that is available in the 
[Bioconductor Project](http://bioconductor.org/).
Bioconductor is similar to the 
[Comprehensive R Archive Network (CRAN)](https://cran.r-project.org/), 
in that you can install packages from this repository.

[`EBImage`](https://bioconductor.org/packages/release/bioc/html/EBImage.html) 
provides general purpose functionality for image processing and analysis. In 
the context of (high-throughput) microscopy-based cellular assays, EBImage 
offers tools to segment cells and extract quantitative cellular descriptors. 
This allows the automation of such tasks using the R programming language and 
facilitates the use of other tools in the R environment for signal processing, 
statistical modeling, machine learning and visualization with image data.

This [chapter in Modern Statistics for Modern Biology](https://web.stanford.edu/class/bios221/book/11-chap.html) is a great reference for using `EBImage` for different types of imaging data.

To install the `EBImage` package, you can run the chunk below. This will 
check whether you have the `BiocManager` package installed, and if not it will
install `BiocManager`. Then, the `EBImage` package will be installed via
the `BiocManager::install()` function.

Note: if you attempt to run `install.packages("EBImage")` you may be met with 
an error! This is because the package is available in Bioconductor and not on 
CRAN.

```
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("EBImage")
```

# Convert cell morphology `.ome.tif` to `.tif` format

The raw data bundle contains the cell morphology image in the [`ome.tif`](https://docs.openmicroscopy.org/ome-model/5.6.3/ome-tiff/) file format. 
This type of file includes the image pixel intensities as well as additional metadata
that is associated with the microscopy experiment. Since we are interested in the
image information only, we need to convert to a `.tif` format to enable 
further downstream processing with `EBImage`.

You are given the `.tif` file in the Processed Data Bundle, but you can see
how this was generated in the rest of this section.

Load the `EBImage` R package.

```{r, setup}
library(EBImage)
```

Read in morphology focus `.ome.tif` image and export out as a `.tif` into the 
`../data/` folder. Only do so if the target `.tif` file does not exist.

Note that if we need to generate the `.tif` file, we need to first set up 
the java memory to 10GB and load the `RBioFormats` package, which is available
on [Bioconductor development branch](https://www.bioconductor.org/packages/devel/bioc/html/RBioFormats.html) and on [Github](https://github.com/aoles/RBioFormats).

```{r}
tifFile = "../morphology_focus.tif"
if (!file.exists(tifFile)) {
  options(java.parameters = "-Xmx10g")
  library(RBioFormats)
  checkJavaMemory()
  img_ome = RBioFormats::read.image("../data_raw/morphology_focus.ome.tif",
                                    read.metadata = FALSE,
                                    normalize = TRUE)
  
  img = img_ome[[1]]@.Data
  EBImage::writeImage(x = img,
                      files = tifFile,
                      type = "tiff")
}
```

# Copy tabular cell data to outside of raw directory

Since the raw data bundle contains many large files, for convenience we have 
copied three files from the `../data_raw/` directory to outside of that directory.
This can be done programmatically using the `system()` function.

```
system("cp ../data_raw/cell_boundaries.csv.gz ../cell_boundaries.csv.gz")
system("cp ../data_raw/analysis/clustering/gene_expression_graphclust/clusters.csv ../clusters.csv")
system("cp ../data_raw/analysis/diffexp/gene_expression_graphclust/differential_expression.csv ../differential_expression.csv")

```

# Finish

```{r}
sessionInfo()
```

```{r}
knitr::knit_exit()
```
